libbeat/publisher/queue/spool/inbroker.go

```diff
@@ -0,0 +1,482 @@
+	// =>
+	//   - keep pointer to yet unwritten event contents
+	//   - do not accept any events if pending is not nil
+	//   - wait for signal from reader/queue-gc to retry writing the pending
+	//     events contents
+	pending []byte
+
+	bufferedEvents uint // number of buffered events
+
+	timer *timer
+	enc   *encoder
+}
+
+func newInBroker(
+	ctx *spoolCtx,
+	qu *pq.Queue,
+	flushTimeout time.Duration,
+	codec codecID,
+) (*inBroker, error) {
+	enc, err := newEncoder(codec)
+	if err != nil {
+		return nil, err
+	}
+
+	b := &inBroker{
+		ctx:   ctx,
+		state: (*inBroker).stateEmpty,
+
+		// API
+		events:    make(chan pushRequest, 20),
+		pubCancel: make(chan producerCancelRequest),
+		sigACK:    make(chan struct{}, 3),
+		sigFlush:  make(chan uint, 3),
+		ackDone:   make(chan struct{}),
+
+		// queue state
+		queue:          qu,
+		clientStates:   clientStates{},
+		pending:        nil,
+		bufferedEvents: 0,
+
+		// internal
+		timer: newTimer(flushTimeout),
+		enc:   enc,
+	}
+
+	ctx.Go(b.eventLoop)
+	ctx.Go(b.ackLoop)
+	return b, nil
+}
+
+func (b *inBroker) Producer(cfg queue.ProducerConfig) queue.Producer {
+	return newProducer(b.ctx, b.pubCancel, b.events, cfg.ACK, cfg.OnDrop, cfg.DropOnCancel)
+}
+
+// onFlush is run whenever the queue flushes it's write buffer. The callback is
+// run in the same go-routine as the Flush was executed from.
+// Only the (*inBroker).eventLoop triggers a flush.
+func (b *inBroker) onFlush(n uint) {
+	if n > 0 {
+		b.ctx.logger.Debug("inbroker: flushed events:", n)
+		b.bufferedEvents -= n
+		b.sigFlush <- n
+	}
+}
+
+// onACK is run whenever the queue releases ACKed events. The number of acked
+// events and freed pages will is reported.
+// Flush events are forward to the brokers eventloop, so to give the broker a
+// chance to retry writing in case it has been blocked on a full queue.
+func (b *inBroker) onACK(events, pages uint) {
+	if pages > 0 {
+		b.sigACK <- struct{}{}
+	}
+}
+
+func (b *inBroker) ackLoop() {
+	log := b.ctx.logger
+
+	log.Debug("start flush ack loop")
+	defer log.Debug("stop flush ack loop")
+
+	for {
+		var n uint
+		select {
+		case <-b.ackDone:
+			return
+
+		case n = <-b.sigFlush:
+			log.Debug("inbroker: receive flush", n)
+			states := b.clientStates.Pop(int(n))
+			b.sendACKs(states)
+		}
+	}
+}
+
+// sendACKs returns the range of ACKed/Flushed events to the individual
+// producers ACK handlers.
+func (b *inBroker) sendACKs(states []clientState) {
+	log := b.ctx.logger
+
+	// reverse iteration on client states, so to report ranges of ACKed events
+	// only once.
+	N := len(states)
+	total := 0
+	for i := N - 1; i != -1; i-- {
+		st := &states[i]
+		if st.state == nil {
+			continue
+		}
+
+		count := (st.seq - st.state.lastACK)
+		if count == 0 || count > math.MaxUint32/2 {
+			// seq number comparison did underflow. This happens only if st.seq has
+			// already been acknowledged
+			// log.Debug("seq number already acked: ", st.seq)
```

Comment:
It's some kind of tracing (we don't have trace level in beats). For debugging locally it's super helpful, but otherwise it's much to noisy for users (every event ever published will trigger a log line).